hs/src/Reach/Eval/Core.hs

```diff
@@ -5465,6 +5479,9 @@ evalStmt = \case
             dv' <- ctxt_mkvar $ DLVar at_c (Just (at_c, de_v)) vt
             let vv = case vt of
                   T_Null -> SLV_Null at_c "case"
+                  T_Token -> case de_val of
+                      SLV_Data _ _ _ (SLV_DLTok (DLToken _ i)) -> SLV_DLTok (DLToken dv' i)
```

Comment:
This line shows the brittleness of this technique. I think if you write more test scenarios, you'll see problems with it. For example... a tuple of two tokens, an array of tokens, a function that returns a token based on some condition, a token sent in to the program that is identical to an earlier token (asserted with a require)

I think you could generalize and improve your technique a little bit by making the `ST_Token` type have a `(Maybe Int)` parameter which is used to identify which it is and if it is `Nothing`, then that means "any token". I'm not sure how our type machinery will work with that, but I hypothesize it will be easier. I don't think that the DL types would change at all, but maybe I'm wrong because of the way balance assertions work.

I think the most powerful technique will be "valued based". Remember, the reason we need to keep track of these things is so that we can (a) track the balance adjustments and (b) track the balances in loops with the fluid variable capture. The dumbest thing that could possibly work would be (1) representing the balances as `Map`-like type in DL, (2) using SMT's maps for that during verification, and (3) compiling it to an array on a platform like ALGO that doesn't have maps (and maybe it would be good to be an array on others when we know the identity of tokens statically.) This would basically be like adding a new `T_TokenBalances` type.

Maybe my two ideas are too heavy, but I don't think we can evaluate your idea without more examples of what it allows and doesn't allow.

Reply:
I get why this seems brittle, but in this case, we know that if the unboxed value being bound by a pattern match is of type `T_Token`, it can only be `SLV_Data _ _ _ (SLV_DLTok (DLToken _ i))`.

I will update the example to demonstrate how it works with arrays, functions etc. so long as it's statically computable. When a token sent in to the program is identical to an earlier token, it errors as expected.

I think the DL `T_Token` type would require the `Maybe (Int)` information for looking up the balances. In my opinion, this approach could get ugly later in the compiler when we have Map keys that are `T_Token`, so we'd probably have to intercept lookups to treat `lookup (T_Token (Just ..)) map` as `lookup (T_Token Nothing) map`

I kind of get the gist of the value-based approach, but I don't think I understand it well. For

> (1) representing the balances as Map-like type in DL

Would Tokens be `DLVar`s with type: `T_TokenBalances (Map DLVar Int)`?

Reply:
Let's break this conversation up into two pieces.

First, I want more systematic and small tests (okay if they are just in `hs/t` and compile-time only) that show what is possible _now_, with *this* implementation, and what is still not possible.

The way I'd describe in words is that today, tokens must appear at the top-level of a publish and must remain completely un-abstracted in every operation, thus they must be compile-time static values the entire time. So, you can't publish a Maybe Token and expect it to be visible, you have to publish the actual token. Inside one consensus step, you can't stick two tokens into a container and then pull them out and then be able to use them afterwards. What about across a consensus step... does that matter?

With this change, I think that token identities are preserved in DLArgs, but I don't have a good sense of what that allows or doesn't.

--

Second, the different implementations.

About the type idea... you have introduced an argument `DLToken DLVar Int` with the invariant that `DLVar` is always `DLVar src lab T_Token id`. My point is just that `DLToken (DLVar src lab T_Token id) which` is equivalent to storing the `which` in the token type `DLVar src lab (T_Token (Just which)) id` and is like the common dependent type concept of an indexed type, so I think it would fit in better if you did it that way. (You need the `Maybe`, because in your version some `DLVar`s that are `T_Token` aren't wrapped in a `DLToken`.)

About the map idea... what's the point of keeping track of tokens? It is so that when you do `transfer(x, tok)` we can do `(assert (> (balance tok) x))` and `(fluid-set (balance tok) (- (balance tok) x)`. Everything about tokens is all about that little snippet. Right now, we do that by implementing `(balance tok)` with a fixed array that is referenced by position which requires us to know which index `tok` is at. It's like its type is `Array(UInt, HowManyTokensInProgram)`. We could instead represent it with `Array(Tuple(Token, UInt), HowManyTokensInProgram)` and then implement (balance tok)` with a simple search. In SMT, we could compile that "search" with an actual map. With some clever compilation, we may also be able to get the exact position a lot of the time. I have an idea for how to implement this smoothly that I can explain also, but I think I'd rather speak it than write it.